### PR TITLE
Add gossip notifications / cluster node notifications to geyser

### PR DESCRIPTION
**Pull request by musitdev**
#### Problem
See issue #32934

#### Summary of Changes
This PR is a POC to add cluster node's info to the Geyzer plugin notification.

The LegacyContactInfo received on GOSSIP network are forwarded with the Geyzer interface to a client subscriber.

The PR contains:
 * the entry point in crds.rs to notify LegacyContactInfo [updates](https://github.com/solana-labs/solana/blob/f4bc1be708298c0896978c1ce286dcdbf1ecc221/gossip/src/crds.rs#L327) and [removals](https://github.com/solana-labs/solana/blob/f4bc1be708298c0896978c1ce286dcdbf1ecc221/gossip/src/crds.rs#L553),
 * the Geyzer plugin interface update to add [LegacyContactInfo notification](https://github.com/solana-labs/solana/blob/f4bc1be708298c0896978c1ce286dcdbf1ecc221/geyser-plugin-interface/src/geyser_plugin_interface.rs#L20) and [notification entry point](https://github.com/solana-labs/solana/blob/f4bc1be708298c0896978c1ce286dcdbf1ecc221/geyser-plugin-interface/src/geyser_plugin_interface.rs#L354).
 * the processing of the updates by the [geyzer_plugin_manager](https://github.com/solana-labs/solana/blob/f4bc1be708298c0896978c1ce286dcdbf1ecc221/geyser-plugin-manager/src/cluster_info_notifier.rs).

A GRPC [Yellowstone plugin update](https://github.com/rpcpool/yellowstone-grpc/pull/268) has been done to test the process and see the notified data volume. You can see the PR here.

An example of client can be found [here](https://github.com/blockworks-foundation/solana-rpc-v2/blob/main/clusterinfo/src/main.rs). It subscribes to the Cluster info node notification and call from time to time the RPC *get_cluster_nodes* to compare with the notified data.

##### Early results

The test has been done on Testnet cluster with the Solana version 1.17.5 and Yellowstone plugin 1.11.0+solana.1.17.5.

###### Comparison with RPC call
It takes a few second for the plugin to get the majority of the Testnet cluster nodes and after 30s there's less than 10 nodes in error when comparing with the RPC call (around 3000 nodes in the cluster).

After a few minutes, Geyzer and RPC has the same list of cluster's nodes. From time to time they can have some difference (a few nodes) from one part or the other depending on the notification time propagation of update or removal.

###### Data volume

I use Tcpdump to monitor the packet send between the validator grpc server and the client to build the *get_cluster_nodes RPC* call nodes list. 

This graphic is an example of network activity and bandwidth use. The volume of data transmitted is between 0,5 Mbits/s and 3 Mbits/s.

I've done some monitoring of the network connection between the grpc server and the client for the cluster node info notification. 

This is the graphic of the data transferred between the validator grpc server and the client to build the *get_cluster_nodes* RPC call nodes list.

Network metrics

### CODE REVIEW

**Comment by willhickey** [comment]
v1.17 is nearly at the end of its stabilization and only accepting PRs for bug fixes. Please consider opening this PR against master instead.

**Comment by musitdev** [comment]
> v1.17 is nearly at the end of its stabilization and only accepting PRs for bug fixes. Please consider opening this PR against master instead.

I've used the 1.17 to have the grpc geyzer plugin compatible in version. I've created this PR first to do a test to validate the concept. It wasn't intended to be merged.

If it seems to you interesting, I can move my change to the current master and try to adapt the grpc plugin to work with it.

File: gossip/src/crds.rs [hunk 1]
```diff
@@ -309,7 +323,10 @@ impl Crds {
                     Err(CrdsError::InsertFailed)
                 }
             }
-        }
+        };
+        //notify geyzer interface
+        self.notify_clusterinfo_update(self.table.get(&gossip_label));
```
Comment (lijunwangs) [inline gossip/src/crds.rs:328]: What about for the upsert case?

File: gossip/src/crds/geyser_plugin_utils.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+use crate::crds::Crds;
+use crate::crds::Pubkey;
+use crate::crds::VersionedCrdsValue;
+use crate::crds_value::CrdsData;
+
+impl Crds {
+    /// Notified when an account is updated at runtime, due to transaction activities
```
Comment (lijunwangs) [inline gossip/src/crds/geyser_plugin_utils.rs:7]: Correct these copy-pasted comments please

File: gossip/src/crds.rs [hunk 3]
```diff
@@ -533,6 +550,8 @@ impl Crds {
         self.shards.remove(index, &value);
         match value.value.data {
             CrdsData::LegacyContactInfo(_) => {
+                //notify geyzer interface
```
Comment (lijunwangs) [inline gossip/src/crds.rs:553]: nits: --> geyser

File: geyser-plugin-interface/src/geyser_plugin_interface.rs [hunk 4]
```diff
@@ -2,6 +2,8 @@
 /// the GeyserPlugin trait to work with the runtime.
 /// In addition, the dynamic library must export a "C" function _create_plugin which
 /// creates the implementation of the plugin.
+use solana_sdk::pubkey::Pubkey;
```
Comment (lijunwangs) [inline geyser-plugin-interface/src/geyser_plugin_interface.rs:5]: Follow our coding style, put them in "use {...}";

File: geyser-plugin-interface/src/geyser_plugin_interface.rs [hunk 5]
```diff
@@ -317,6 +349,18 @@ pub trait GeyserPlugin: Any + Send + Sync + std::fmt::Debug {
...
 
+    /// Called when a cluster info is updated on gossip network.
+    #[allow(unused_variables)]
+    fn update_cluster_info(&self, cluster_info: &ReplicaClusterInfoNode) -> Result<()> {
+        Ok(())
+    }
+
+    /// Called when a cluster info is removed on gossip network.
+    #[allow(unused_variables)]
+    fn notify_clusterinfo_remove(&self, pubkey: &Pubkey) -> Result<()> {
```
Comment (lijunwangs) [inline geyser-plugin-interface/src/geyser_plugin_interface.rs:360]: When is this called? notify_clusterinfo_remove?

**Review by lijunwangs — COMMENTED** [review]
Please move the PR to against the master branch.

File: gossip/src/crds.rs (hunk 1 above)
Comment (musitdev) [reply gossip/src/crds.rs:328]: Sorry, I didn't understand. It's clusterinfo you prefer cluster_info?

File: gossip/src/crds.rs (hunk 3 above)
Comment (musitdev) [reply gossip/src/crds.rs:553]: Corrected

File: geyser-plugin-interface/src/geyser_plugin_interface.rs (hunk 4 above)
Comment (musitdev) [reply geyser-plugin-interface/src/geyser_plugin_interface.rs:5]: Corrected

File: geyser-plugin-interface/src/geyser_plugin_interface.rs (hunk 5 above)
Comment (musitdev) [reply geyser-plugin-interface/src/geyser_plugin_interface.rs:360]: You can find the call here: https://github.com/solana-labs/solana/blob/f4bc1be708298c0896978c1ce286dcdbf1ecc221/gossip/src/crds.rs#L554

**Comment by musitdev** [comment]
I've created a new PR to start from master. I've issues with master rebase so I apply the commit to master branch directly and use a new branch.
The new PR is #32934

**Pull request closed without merging**